mtags/src/main/scala-3/scala/meta/internal/pc/completions/MatchCaseCompletions.scala

```diff
@@ -0,0 +1,479 @@
+    val clientSupportsSnippets = config.isCompletionSnippetsEnabled()
+    lazy val autoImportsGen = AutoImports.generator(
+      pos,
+      text,
+      typedTree,
+      indexedContext,
+      config,
+    )
+    val completionGenerator = CompletionValueGenerator(
+      indexedContext,
+      completionPos,
+      clientSupportsSnippets,
+      patternOnly,
+      hasBind,
+    )
+    val parents: Parents = selector match
+      case EmptyTree =>
+        val seenFromType = parent match
+          case TreeApply(fun, _) if fun.tpe != null && !fun.tpe.isErroneous =>
+            fun.tpe
+          case _ =>
+            parent.tpe
+        seenFromType.paramInfoss match
+          case (head :: Nil) :: _
+              if definitions.isFunctionType(head) || head.isRef(
+                definitions.PartialFunctionClass
+              ) =>
+            val dealiased = head.widenDealias
+            val argTypes =
+              head.argTypes.init
+            new Parents(argTypes, definitions)
+          case _ =>
+            new Parents(NoType, definitions)
+      case sel => new Parents(sel.tpe, definitions)
+
+    val result = ListBuffer.empty[CompletionValue]
+    val isVisited = mutable.Set.empty[Symbol]
+    def visit(sym: Symbol, name: String, autoImports: List[l.TextEdit]): Unit =
+      val isValid = !isVisited(sym) && !parents.isParent(sym)
+        && (sym.is(Case) || sym.is(Flags.Module) || sym.isClass)
+        && parents.isSubClass(sym, false)
+      def recordVisit(s: Symbol): Unit =
+        if s != NoSymbol && !isVisited(s) then
+          isVisited += s
+          recordVisit(s.moduleClass)
+          recordVisit(s.sourceModule)
+      if isValid then
+        recordVisit(sym)
+        if completionGenerator.fuzzyMatches(name) then
+          result += completionGenerator.toCompletionValue(
+            sym,
+            name,
+            autoImports,
+          )
+        end if
+      end if
+    end visit
+    val selectorSym = parents.selector.typeSymbol
+
+    // Special handle case when selector is a tuple or `FunctionN`.
+    if definitions.isTupleClass(selectorSym) || definitions.isFunctionClass(
+        selectorSym
+      )
+    then
+      val label =
+        if patternOnly.isEmpty then s"case ${parents.selector.show} =>"
+        else parents.selector.show
+      result += CompletionValue.CaseKeyword(
+        selectorSym,
+        label,
+        Some(
+          if patternOnly.isEmpty then
+            if config.isCompletionSnippetsEnabled() then "case ($0) =>"
+            else "case () =>"
+          else if config.isCompletionSnippetsEnabled() then "($0)"
+          else "()"
+        ),
+        Nil,
+        range = Some(completionPos.toEditRange),
+        command = config.parameterHintsCommand().asScala,
+      )
+    else
+      // Step 1: walk through scope members.
+      indexedContext.scopeSymbols
+        .foreach(s => visit(s.info.dealias.typeSymbol, s.decodedName, Nil))
```

Comment:
As I got it was added as there were issues with some particular cases like for `Nil$` and `Nil` in `indexedContext.findSymbol`.

I started looking if it's possible to fix `IndexedContext` to avoid checking all symbols in scope.

Reply:
There was already a discussion about going through scope symbols here https://github.com/scalameta/metals/pull/4229#discussion_r944422269
It's not because of a bug, but for not sealed traits. Actually I already have a prepared improvement, by moving the `isValid` filter from `visit` to something like `indexedContext.scopeSymbols.filter(isValid).foreach(..)`, but I wanted to do it in a follow up to not change anything more in this pr

Reply:
Or maybe we can use `indexedContext.scopeSymbols` as a fallback when `sealedStrictDescendants` is empty?

Reply:
@jkciesluk thx for pointing at discussion. yep - let's solve it after merging this one